cgp/genome.py

```diff
@@ -267,6 +267,181 @@ def randomize(self, rng: np.random.RandomState) -> None:
+                gene_idx
+            ):
+                try:
+                    gene_value = self._n_inputs + used_idx_list.index(gene_value)
+                except ValueError:  # means the input is from an input node -> do nothing
+                    pass
+            dna_for_reorder[gene_idx] = gene_value
+
+    def _replace_invalid_gene_values(
+        self, dna_for_reorder: List[int], rng: np.random.RandomState
+    ) -> None:
+        """ Replace gene values of unused arities
+        by (pseudo-) random picks of permissible values for that node
+        """
+        for gene_idx, gene_value in enumerate(dna_for_reorder):
+            region_idx = gene_idx // self._length_per_region
+            if self._is_hidden_input_gene(gene_idx, region_idx) and gene_value > region_idx:
+                # replace value that is too large by smaller value
+                permissible_values = self.determine_permissible_values_per_gene(gene_idx)
+                gene_value = rng.choice(permissible_values)
+                dna_for_reorder[gene_idx] = gene_value
+
+    def _determine_dependencies(self) -> Dict:
+        """ Determines for every node a set of dependencies.
+
+            The set of dependencies for a node is given by the indices of the nodes,
+            which explicitly must occur earlier in the genome.
+            The arity of the operator gene is taken into account, only nodes which
+            are used by the operator are seen as dependency.
+            Implicit dependencies are not added, since they are covered via the
+            set of explicit dependencies of the respective node.
+
+
+        Returns
+        ----
+        dependencies: Dict
+            Dictionary containing for every node a set of internal nodes dependencies
+
+        """
+        dependencies = {}
+        for region_idx in range(self._n_regions):
+            current_node_dependencies: set = set()
+
+            if self._is_input_region(region_idx) or self._is_output_region(region_idx):
+                continue
+
+            elif self._is_hidden_region(region_idx):
+
+                operator_idx = region_idx * self._length_per_region
+
+                # only consider genes which are used in current node
+                current_arity = self._determine_operator_arity(operator_idx)
+
+                # append each input to the dependencies
+                for idx_gene in range(
+                    1, current_arity + 1
+                ):  # shift by 1 since first gene is the operator gene
+                    input_node_idx = self._dna[operator_idx + idx_gene]
+                    if not self._is_input_region(
+                        input_node_idx
+                    ):  # not necessary to add input regions, since they are first anyway
+                        current_node_dependencies.add(input_node_idx)  # append the input node &
+
+                dependencies[str(region_idx)] = current_node_dependencies
```

Comment:
why `str`?

Reply:
i thought only `str` could be used as keys in dict. Removed it.